internals/overlord/checkstate/manager.go

```diff
@@ -58,13 +59,20 @@ func (m *CheckManager) PlanChanged(p *plan.Plan) {
 	for _, check := range m.checks {
 		check.cancel()
 	}
+	// Wait for all context cancellations to propagate and allow
+	// each goroutine to cleanly exit.
+	m.group.Wait()
+
+	// Set the size of the next wait group
+	m.group.Add(len(p.Checks))
 
 	// Then configure and start new checks.
 	checks := make(map[string]*checkData, len(p.Checks))
 	for name, config := range p.Checks {
 		ctx, cancel := context.WithCancel(context.Background())
 		check := &checkData{
 			config:  config,
+			group:   &m.group,
```

Comment:
I think it's a bit simpler to keep the wait group out of the `checkData` struct, so the check itself doesn't have to care about it and all the `WaitGroup` stuff is managed here. You'd do this by saying:

```go
for name, config := range p.Checks {
    ctx, cancel := context.WithCancel(context.Background())
    check := &checkData{...} // no "group" in here anymore
    checks[name] = check
    go func() {
        defer m.group.Done()
        check.loop()
    }()
}
```

Reply:
Wow, crazy cool. Makes a lot of sense.